cmd/k8s-operator: allow to install operator via helm

irbekrm opened the pull request:
Initial helm manifests with basic configurability.

To try this out:

1. run `helm package ./cmd/k8s-operator/deploy/chart` (users won't have to do this once we have published charts)
2. either create a `Secret` with [operator oauth creds](https://tailscale.com/kb/1236/kubernetes-operator/?q=kubernetes#setting-up-the-kubernetes-operator) or put the creds in ./cmd/k8s-operator/deploy/chart/values.yaml (or create a new values file)
3. (if required) edit any other values you want to change in values file
4. run `helm install operator -n tailscale -f <path-to-values-file> tailscale-operator-0.1.0.tgz`
5. observe the operator be successfully deployed in tailscale namespace

The next step will be to set up some basic infra to serve the chart as well as a flow to publish new chart versions.
Updates tailscale/tailscale#9222

There will definitely be lots of additional configuration options that we'll want to add- this PR does not aim to cover them all, but rather to get the initial template in so that we have something.

irbekrm commented on cmd/k8s-operator/deploy/chart/values.yaml:
```diff
@@ -0,0 +1,36 @@
+  hostname: "tailscale-operator"
+  nodeSelector:
+    kubernetes.io/os: linux
+  resources: {}
```
I took out the resource request that we had. I think it does not make sense to have these values here- the actual requests will depend on the amount and size of resources that the operator needs to parse, so users would  have to figure it out for themselves anyway. These requests were a lot higher than a user who just wants to set up a couple proxies would need and were making the deployment to fail on clusters that don't have a large amount of  free resources.
For reference bitnami charts, also don't preset resource requests/limits on their charts https://github.com/bitnami/charts/blob/main/bitnami/cert-manager/values.yaml#L129-L131

maisem commented on cmd/k8s-operator/deploy/chart/templates/deployment.yaml:
```diff
@@ -0,0 +1,86 @@
+          {{- end }}
+          image: {{ .Values.operatorConfig.image.repo }}:{{ .Values.operatorConfig.image.tag | default .Chart.AppVersion }}
+          imagePullPolicy: {{ .Values.operatorConfig.image.pullPolicy }}
+          resources:
+            {{- toYaml .Values.operatorConfig.resources | nindent 12 }}
```
do the same `{{- with .Values.operatorConfig.resources }}`. it omits it completely from generated yaml

irbekrm replied:
nice

maisem commented on tailscale-operator/.helmignore:
```diff
@@ -0,0 +1,23 @@
+# Patterns to ignore when building packages.
+# This supports shell glob matching, relative path matching, and
+# negation (prefixed with !). Only one pattern per line.
+.DS_Store
+# Common VCS dirs
+.git/
```
move this file too

maisem commented on cmd/k8s-operator/deploy/manifests/operator.yaml:
we would need to update docs

irbekrm replied:
will do after this merges

mayakacz commented:
Updates ENG-1837

maisem commented on cmd/k8s-operator/deploy/chart/templates/oauth-secret.yaml:
```diff
@@ -0,0 +1,12 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: operator-oauth
```
namespace

irbekrm replied:
Ahh I didn't realize we actually set namespace on each object. We didn't need to, it automatically adds namespace passed via `--namespace` flag to namespaced resources

irbekrm replied:
However, I also don't see any harm from adding it- will add for consistency

maisem commented on cmd/k8s-operator/deploy/chart/Chart.yaml:
```diff
@@ -0,0 +1,27 @@
+# This is the chart version. This version number should be incremented each time you make changes
+# to the chart and its templates, including the app version.
+# Versions are expected to follow Semantic Versioning (https://semver.org/)
+version: 0.1.0
+
+# This is the version number of the application being deployed. This version number should be
+# incremented each time you make changes to the application. Versions are not expected to
+# follow Semantic Versioning. They should reflect the version the application is using.
+# It is recommended to use it with quotes.
+appVersion: "unstable"
```
@irbekrm @mayakacz @rhealitycheck  any thoughts on how we version these?
Should we just follow the tailscale versioning?

Stelminator replied:
@maisem 
https://helm.sh/docs/topics/charts/
> version: A SemVer 2 version (required)
> appVersion: The version of the app that this contains (optional). Needn't be SemVer. Quotes recommended.

`appVersion` should likely match the latest stable operator version, so `v1.50.1` for the container now. If this changes, bump the patch component of `version`

`version` is SemVer for the *chart itself*. i.e., bump the main version if you're breaking backward compatibility. Bump minor/patch when making any change to the chart. 0.1.0 would be appropriate for the first merge (0.x.y indicating that it's not really done). This is independent of `appVersion`. For example, if the entire chart got refactored, that's probably a major version change, but it wouldn't be a tailscale major version. Conversely, there might not be any structural changes to the chart for a very long time, making it forward compatible with tailscale `v3.1.2`.

Here's a couple of examples:
- https://github.com/grafana/helm-charts/pull/2716/files (chart added a field, patch++)
- https://github.com/grafana/helm-charts/pull/2706/files (appVersion bump, patch++)

irbekrm replied:
> Should we just follow the tailscale versioning?

Yes, previously I have bumped this every time I release a new chart version to make sure the right images get pulled.
This is also the tag of the operator and proxy images so I've set this to `unstable` here because that's what we currently have in the plain manifests.

It doesn't really matter what versions are there _in this PR_ because we aren't actually going to publish with these versions.Once we set up infra to actually publish charts we'll have to have a discussion about publishing flow and versioning. And before we publish we'll have to bump these anyway.

irbekrm replied:
wrt, to the chart version, I am not sure whether we are likely to be publishing new chart versions outside of Tailscale release flow. If we do not, then having the same app and chart version is fine. We did that for cert-manager for example.

irbekrm replied:
Actually, thinking about it, we shouldn't need to ever bump it here. It should be ok that both `appVersion` and `version` are dummy values and we set them at publish time `helm package <path> --app-version v0.52.0 --version v0.52.0`.
See i.e https://github.com/cert-manager/cert-manager/blob/master/deploy/charts/cert-manager/Chart.template.yaml#L3-L5
I've set the proxy tag in values.yaml to `""` to ensure that `unstable` does not override `appVersion` that we default to.

irbekrm commented on cmd/k8s-operator/deploy/chart/Chart.yaml:
```diff
@@ -0,0 +1,29 @@
+
+type: application
+
+maintainers:
```
Not sure if we have some kind of 'maintainers' email address that we could put here

irbekrm commented:
> [irbekrm](https://github.com/irbekrm) [force-pushed](https://github.com/tailscale/tailscale/compare/ced02933b7a9ce237983b2de822f661d240daa99..895a5ce2af256d3b11f7f012cc6933d09b948164) the irbekrm/helmchart branch from [ced0293 ](https://github.com/tailscale/tailscale/commit/ced02933b7a9ce237983b2de822f661d240daa99)to [895a5ce ](https://github.com/tailscale/tailscale/commit/895a5ce2af256d3b11f7f012cc6933d09b948164)
[2 minutes ago](https://github.com/tailscale/tailscale/pull/9920#event-10808340396)

Added ability to specify a digest for operator/proxy images and added a couple things (keywords etc) to Chart.yaml
